timon/run.py

```diff
@@ -24,16 +24,28 @@
 logger = logging.getLogger(__name__)
 
 
-async def ask_exit(signame, cfg):
+async def ask_exit(signame, stopevent, cfg):
     """
     This code is an attempt to help performing a
-    clean shutdown
-    At moment, only stop plugins and stop the dbstore
+    clean shutdown.
+    First time the signal is catched, set an asyncio event to
+    True that permits to not relaunch the loop and cleanly stop after
+    probes process.
+    Second time the signal is catched, abruptly stop timon (
+    but cleanly stop plugins and the dbstore to avoid loosing data)
     """
-    logger.info("got signal %r: will exit", signame)
-    await cfg.stop_plugins()
-    cfg.stop_dbstore()
-    sys.exit()
+    logger.info("got signal %r: will ask exit", signame)
+    if not stopevent.is_set():
+        stopevent.set()
+    else:
+        logger.warning("Second stop signal: Force exit")
+        try:
+            async with asyncio.timeout(5):
+                await cfg.stop_plugins()
+                cfg.stop_dbstore()
```

Comment:
cacn close. so stop_dbstore is syncrhonous?

Reply:
Yup, dbstore is a thread and the stop_func is synchronous